### PR TITLE
Diagnostics tools - Network throttling

### PR DESCRIPTION
**Pull request by aryszka**
This pull request implements network throttling: latency, max bandwidth or custom chunking with delays.

### CODE REVIEW

**Comment by mo-gr** [comment]
👍

**Comment by szuecs** [comment]
:+1:

File: filters/diag/diag.go [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+func NewRandom() filters.Spec { return &random{} }
+
+// NewLatency creates a filter specification whose filter instances can be used
+// to add additional latency to responses. It expects the latency in milliseconds
+// as an argument. It always adds this value in addition to the natural latency,
+// and does not do any adjustments. Eskip example:
+//
+// 	* -> latency(120) -> "https://www.example.org";
+//
+func NewLatency() filters.Spec { return &throttle{typ: latency} }
```
Comment (lmineiro) [inline filters/diag/diag.go]: I would've preferred that all the latency arguments to be strings and accept valid ParseDuration values. We could still accept simple numbers and default that to ms. It's a matter of taste but I'd like to see arguments such as "1s", "500µs", etc.
One thing to filter out, just like right now, would be negative latencies.

File: filters/diag/diag.go [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &throttle{t.typ, chunkSize, delay}, nil
+}
+
+func (t *throttle) goThrottle(in io.ReadCloser, close bool) io.ReadCloser {
+	if t.chunkSize <= 0 {
+		if t.delay > 0 {
```
Comment (lmineiro) [inline filters/diag/diag.go]: this is not needed: "A negative or zero duration causes Sleep to return immediately."

File: filters/diag/diag.go [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+		defer func() {
+			w.CloseWithError(err)
+			if close {
+				in.Close()
+			}
+		}()
+
+		b := make([]byte, defaultChunkSize)
+		for err == nil {
+			n := 0
```
Comment (lmineiro) [inline filters/diag/diag.go]: Should we be specific about `uint64` or at least some other kind of `uint`?

File: filters/diag/diag.go [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+
+			if err == nil {
+				delay := t.delay
+
+				switch t.typ {
+				case bandwidth, backendBandwidth:
+					delay -= time.Now().Sub(start)
+				}
+
+				if delay >= 0 {
```
Comment (lmineiro) [inline filters/diag/diag.go]: It's not needed

File: filters/diag/diag_test.go [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+func checkMessage(expect messageExp, start time.Time, body io.Reader) error {
+	if expect.header > 0 && !checkWithTolerance(start, expect.header) {
+		return fmt.Errorf("expected time to header, %v, %v, %v", start, time.Now(), expect.header)
+	}
+
+	totalRead := 0
+
+	cstart := start
+	for _, c := range expect.chunks {
+		n := 0
```
Comment (lmineiro) [inline filters/diag/diag_test.go]: Same remark about the data type as before

**Comment by lmineiro** [comment]
This is cool stuff.
I added some minor remarks, none of them blockers for me.

**Pull request merged**